binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+
+## The arrow package
+
+In this chapter we'll rely on Apache Arrow ([arrow.apache.org](https://arrow.apache.org/)), a multi-language toolbox designed for efficient analysis and transport of large data sets.
+The arrow package provides a standard way to use Apache Arrow in R, and allows you to analyze larger-than-memory datasets using familiar dplyr syntax.
+As an additional benefit, arrow is extremely fast: you'll see some examples later in the chapter.
+
+### Opening a dataset
+
+Let's start by taking a look at the Seattle library checkout data.
+At 9GB in size, the csv file is large enough that we probably don't want to load the whole thing into memory.
+Instead of using `read_csv()` we'll use the `open_dataset()` function from the arrow package:
+
+The data set is large enough that we might not want to try loading the entire thing into memory.
+Instead, we'll use `arrow::open_dataset()` to scan the file on-disk and create an Arrow Dataset object called `seattle_lib`:
+
+```{r open-dataset}
+seattle_lib <- open_dataset(
+  sources = "data/seattle-library-checkouts.csv", 
+  format = "csv"
+)
+```
+
+When this code is run, `open_dataset()` scans the file to work out its structure and creates the `seattle_lib` object that stores metadata.
+The data remain on-disk, and are not loaded into memory.
+Nevertheless, we are able to call dplyr functions to interact with the data:
+
+```{r glimpse-data}
+#| cache: true
+glimpse(seattle_lib)
+```
+
+The first line in the output tell you that `seattle_lib` is stored locally on-disk as a single csv file.
+The second line indicates that the data are stored as a tabular data set with 41 million rows and 12 columns.
+
+### Using dplyr with arrow
+
+Writing dplyr code for arrow data is conceptually similar to dbplyr, described in @sec-import-databases.
+The dplyr code that you write in R is internally transformed to a query that the Apache Arrow C++ library understands, which is then executed when you call `collect()`.
+Here's an example that counts the total number of books checked out in each year:
+
+```{r books-by-year}
+#| cache: true
+seattle_lib |> 
+  filter(MaterialType == "BOOK") |>
+  group_by(CheckoutYear) |>
+  summarise(TotalCheckouts = sum(Checkouts)) |>
+  arrange(desc(CheckoutYear)) |>
+  collect()
+```
+
+Only some R expressions are understood by arrow, so you will not necessarily be able to write exactly the same code you usually would.
+However, the list of operations and functions supported is fairly extensive and growing.
+A complete list is provided in the arrow package documentation at [arrow.apache.org/docs/r/reference/acero.html](https://arrow.apache.org/docs/r/reference/acero.html).
+
+## The parquet file format
+
+When working with small data sets, csv files can be convenient: they're human readable plain text files, and most analysis tools can read data from a csv.
+However, there are some drawbacks:
+
+-   Ambiguity.
+    A csv file does not uniquely specify the data structure.
```

Comment:
Need to link to "column types" section (currently only available in https://github.com/hadley/r4ds/pull/1128)